adagrams/game.py

```diff
@@ -1,11 +1,148 @@
+    'W': 4, 
+    'X': 8, 
+    'Y': 4, 
+    'Z': 10 
+}
+
 def draw_letters():
-    pass
+    '''
+    (1-1) create letter bank from LETTER_POOL
+    (1-2) letter cannot repeat greater than quantity of LETTER_POOL
+    '''
+
+    letters = []
+    while len(letters) < 10:
+        letter = random.choice(list(LETTER_POOL.keys()))
+        if letters.count(letter) < LETTER_POOL[letter]:
+            letters.append(letter)
+    return letters
 
 def uses_available_letters(word, letter_bank):
-    pass
+    '''
+    (1) read letter from word (in list) from user
+    (2) read letter_bank from the output of draw_letters()
+    (3) check if each letter in the letter_bank
+    '''
+
+    letters = letter_bank.copy()
+    word = word.upper()
+    for letter in word:
+        if letter in letters:
+            letters.remove(letter)
+            continue
+        else:
+            return False
+    return True
 
 def score_word(word):
-    pass
+    '''
+    (1) letter in word
+    (2) get the value from LETTER_VALUES (in dict)
+    (3) add each value to score
+    '''
+
+    score = 0
+    word = word.upper()
+    for letter in word:
+        score += LETTER_VALUES[letter]
+        
+    if len(word) > 6:
+        score += 8
+
+    return score
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    '''
+    (1) dictionay of word:score
+    (2) get word(s) in dict with highest score
+    (3) convert (2) into tuple (requirement!)
+    (4) list of multiple max scored words
+    (5-1) if (4) is single: return the tuple of max scored word   
+    (5-2) if (4) is multiple: return winner
+    (5-2-a) winner rule for multiple: with fewest number of letters unless 10 letters 
+                                    or if words have same number, then return first occurrence
+    '''
+
+    words_with_scores = {}
+    for word in word_list:
+        score = score_word(word)
+        words_with_scores[word] = score
+    
+    highest_score_word = max(words_with_scores,key=words_with_scores.get) 
+    high_score = words_with_scores[highest_score_word]
+
+    max_score_tuple = (highest_score_word, words_with_scores[highest_score_word])
+    
+    multiple_max_score_words = []
+    for key in words_with_scores:
+        if high_score == words_with_scores[key]:
+            multiple_max_score_words.append(key)
+
+    if len(multiple_max_score_words) == 1:
+        return max_score_tuple
+    else:
+        for i in range(0, len(multiple_max_score_words)):
+            if len(multiple_max_score_words[i]) == 10:
+                return (multiple_max_score_words[i], words_with_scores[multiple_max_score_words[i]])
+            elif len(multiple_max_score_words[i-1]) < len(multiple_max_score_words[i]):
+                current_winner = (multiple_max_score_words[i-1], words_with_scores[multiple_max_score_words[i-1]]) 
+            else:
+                current_winner = (multiple_max_score_words[i], words_with_scores[multiple_max_score_words[i]])
```

Comment:
These two statements do not check what `current_winner` is before changing it. This logic works with the tests we provided, but would fail with other input. 

Say there are three items in `multiple_max_score_words`. Word1 has a length of 2, Word2 has a length of 7, and Word3 has a length of 6.
On the first run of the for loop, Word3's length is NOT shorter than Word1's length, so `current_winner` is set to Word1.
On the second run of the for loop, Word1's length is shorter than Word2's length, so Word1 is set as the `current_winner`.
On the third run, Word2's length is NOT shorter than Word3's length, so Word3 is set as the `current_winner`.
This is the end of the loop, so you return Word3 as the `current_winner`, even though Word1 should be returned.